### PR TITLE
SIMD-0089: Programify Feature Gate Program

### PR DESCRIPTION
**Pull request by buffalojoec**
This proposal suggests replacing the non-existent native program at address
`Feature111111111111111111111111111111111111` with a Core BPF Program, as
described in
[SIMD 0088](https://github.com/solana-foundation/solana-improvement-documents/pull/88).

Feature accounts are already assigned the owner program address
`Feature111111111111111111111111111111111111`. Deploying a Core BPF program at
this address would provide engineers with the capability to revoke pending
feature activations.

**Note:** The process by which core contributors *activate* features would
remain completely unchanged.

### CODE REVIEW

File: proposals/0089-programify-feature-gate-program.md
Comment (CriesofCarrots) [inline proposals/0089-programify-feature-gate-program.md:51]: It's not obvious (to me) how upgrading the program is dangerous, per se. I think I'd just remove that part of this sentence.

File: proposals/0089-programify-feature-gate-program.md
Comment (CriesofCarrots) [inline proposals/0089-programify-feature-gate-program.md:54]:
```suggestion
With the Feature Gate program instead implemented as a Core BPF program, any changes
need only be done once, eliminating this duplication of work.
```

File: proposals/0089-programify-feature-gate-program.md
Comment (CriesofCarrots) [inline proposals/0089-programify-feature-gate-program.md:68]:
```suggestion
will be a multi-sig authority, with keyholders from each validator
```

File: proposals/0089-programify-feature-gate-program.md
Comment (CriesofCarrots) [inline proposals/0089-programify-feature-gate-program.md:74]:
```suggestion
may support in the future are outside the scope of this SIMD.
```
or... "will be proposed and discussed separately"?

File: proposals/0089-programify-feature-gate-program.md
Comment (CriesofCarrots) [inline proposals/0089-programify-feature-gate-program.md:62]:
```suggestion
- "Revoke" or "revoke pending activation": The act of removing a feature from the runtime's recognized set
```
These are design details, so just wait and include below?

File: proposals/0089-programify-feature-gate-program.md
Comment (CriesofCarrots) [inline proposals/0089-programify-feature-gate-program.md:76]:
```suggestion
When this instruction is invoked by
```

File: proposals/0089-programify-feature-gate-program.md
Comment (CriesofCarrots) [inline proposals/0089-programify-feature-gate-program.md:100]: I don't really see much point in this being here. Since we generally try to avoid code in SIMDs, I think just remove it.

File: proposals/0089-programify-feature-gate-program.md
Comment (CriesofCarrots) [inline proposals/0089-programify-feature-gate-program.md:118]:
```suggestion
The official processes
outlined in
[SIMD 0088](https://github.com/solana-foundation/solana-improvement-documents/pull/88)
for migrating a native program to Core BPF will be used to enable this new program, with the addition of the Revoke instruction as a separate BPF rogram-upgrade step.

1. Migrate the native no-op program at
   `Feature111111111111111111111111111111111111` to a Core BPF no-op, with the new program's upgrade authority set to the Feature Gate multisig
2. Upgrade the Core BPF no-op to add the `RevokePendingActivation` instruction using the new multisig authorization

Because the only change to the program is the Revoke addition, these steps will enable BPF Feature Gate without any
changes to existing feature-activation processes.
```

File: proposals/0089-programify-feature-gate-program.md
Comment (buffalojoec) [reply proposals/0089-programify-feature-gate-program.md:100]: Sorry, I thought you guys specifically said to include state and instruction layouts in the SIMDs.

Comment (buffalojoec) [reply proposals/0089-programify-feature-gate-program.md:95]: Circling back here, I plucked the code snippet as requested!

**Comment by buffalojoec** [comment]
Ok @CriesofCarrots I believe I've address all your feedback!

@lheeger-jump @ripatel-fd I would really appreciate you guys having a look at this one as well, when you have a chance.

**Comment by buffalojoec** [comment]
@ripatel-fd I appreciate the thoughts on the larger feature process overhaul, but I think it might be a bit easier to handle the building out of the new process piecemeal, focusing on one step at a time and what each step aims to solve.

This proposal specifically aims to solve the following problem: In our current feature activation process, we want to be able to revoke pending feature activations.

The proposal here is suggesting simply to replace `Feature111111111111111111111111111111111111` with a BPF program that has a revoke instruction. We just need a good way to manage that program's upgrade authority.

> It doesn't seem appropriate to me for a multi-sig to take on the critical task of managing the feature activation queue

This isn't what's being proposed here. Instead this multi-sig would only control the upgrade of the Feature Gate program at `Feature111111111111111111111111111111111111`. The actual control of queued features is unchanged from the current process.

> IMHO, Feature flag activations and core BPF programs should instead only be upgradable via validator governance

I generally agree with this but unfortunately until we have proper validator governance we can't be blocked on engineering efforts. In my opinion a multi-sig is appropriate as an interim control until someday we decide to potentially replace that control with true validator governance, when the mechanism is available.

I think the rest of your thoughts on the larger effort are great and belong in the later SIMDs that will follow this one. What do you think?

File: proposals/0089-programify-feature-gate-program.md
Comment (buffalojoec) [reply proposals/0089-programify-feature-gate-program.md:65]: Ok so this would just be Solana Labs and Jito. I won't go into specifics on how many keys, etc.

I've updated this in the SIMD to remove Jump.

**Review by CriesofCarrots — APPROVED** [review]

**Review by lheeger-jump — APPROVED** [review]
I am going to approve this but I _expect_ it will be amended after we (FD) see a first implementation with more details. Before or after merging, can you link to a reference implementation?

**Comment by buffalojoec** [comment]
> I am going to approve this but I _expect_ it will be amended after we (FD) see a first implementation with more details. Before or after merging, can you link to a reference implementation?

https://github.com/solana-labs/solana-program-library/tree/master/feature-gate

**Comment by lheeger-jump** [comment]
> > I am going to approve this but I _expect_ it will be amended after we (FD) see a first implementation with more details. Before or after merging, can you link to a reference implementation?
> 
> https://github.com/solana-labs/solana-program-library/tree/master/feature-gate
> 
> Keep in mind the actual implementation of the program's source code is decoupled from this SIMD. The proposal herein is outlining:
> 
> * This program will exist as Core BPF
> * This program will be upgradeable, managed by a multi-sig
> * This program will possess the capability to revoke pending feature activations by clearing their data and assigning them back to the system program
> 
> Anything more code-specific beyond that last bullet I believe can be addressed within the PR review process without needing to amend this SIMD, right?

I think so!

**Comment by lheeger-jump** [comment]
Lets merge this on 08/01/2024?

**Comment by buffalojoec** [comment]
> Lets merge this on 08/01/2024?

Sorry, is this DD/MM/YYY?

Fwiw, we need #88 first, but I think we're close to consensus on that one, too.

**Comment by jacobcreech** [comment]
Looks like we've got general consensus. Last call for any objections. We'll merge this after #88 on Thursday, January 25th if no objections are made.

**Review by jacobcreech — APPROVED** [review]
With #88 merged and the last call ended for this one, lgtm. Thanks for pushing this through and gathering consensus

**Pull request merged**
